File: data/walkccc-1666132586/C++/2406 Divide Intervals Into Minimum Number of Groups.cpp
class Solution {
 public:
  // Very simliar to 253. Meeting Rooms II
  int minGroups(vector<vector<int>>& intervals) {
    // Store `right`s
    priority_queue<int, vector<int>, greater<>> minHeap;

    sort(begin(intervals), end(intervals));

    for (const vector<int>& interval : intervals) {
      if (!minHeap.empty() && interval[0] > minHeap.top())
        minHeap.pop();  // No overlap, we can reuse the same group
      minHeap.push(interval[1]);
    }

    return minHeap.size();
  }
};

File: data/walkccc-1666132586/Python/2406 Divide Intervals Into Minimum Number of Groups.py
class Solution:
  # Very simliar to 253. Meeting Rooms II
  def minGroups(self, intervals: List[List[int]]) -> int:
    minHeap = []  # Store `right`s

    for left, right in sorted(intervals):
      if minHeap and left > minHeap[0]:
        heapq.heappop(minHeap)
      heapq.heappush(minHeap, right)

    return len(minHeap)